---
title: "Statistics using R"
author: "Hannah Tavalire and Bill Cresko - University of Oregon"
date: "January 2019 - Cesky Krumlov"
output: ioslides_presentation
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

# But first a beautiful chair

## {.flexbox .vcenter}

```{r, echo=FALSE, out.width='80%', fig.align='center'}
knitr::include_graphics("images/week_01.001.jpeg")
```

## {.flexbox .vcenter}

```{r, echo=FALSE, out.width='100%', fig.align='center'}
knitr::include_graphics("images/week_01.002.jpeg")
```

## {.flexbox .vcenter}

```{r, echo=FALSE, out.width='100%', fig.align='center'}
knitr::include_graphics("images/week_01.003.jpeg")
```

## Goals of this short course in R {.flexbox .vcenter}

- To help you perform your research through instruction in the core components of 
    - data collection and organization
    - manipulation and analysis
    - interpretation and presentation
- Provide a broad coverage of the core components of modern biological statistics
- Provide you with the computational tools necessary to carry out your work - namely R and affiliated tools 

## Why do we need statistics? {.flexbox .vcenter}

- We almost never know the world perfectly, but still want to draw conclusions or make decisions
- We may need to 
  - estimate underlying parameters from samples of data
  - test hypotheses using data
  - summarize and/or visualize large amounts of data
- There are well known mathematical rules that help us
- Statistics can be done by hand, but computers let us do most of the mathematics quickly

## A biological example to get us started {.flexbox .vcenter}

Say you perform an experiment on two different strains of stickleback fish, one from an ocean population (RS) and one from a freshwater lake (BP) by making them microbe free. Microbes in the gut are known to interact with the gut epithelium in ways that lead to a proper maturation of the immune system. 

## A biological example to get us started {.smaller .flexbox .vcenter}

EXPERIMENTAL SETUP - You carry out an experiment by treating multiple fish from each strain so that some of them have a conventional microbiota, and some are inoculated with only one bacterial species. You then measure the levels of gene expression in the stickleback gut using RNA-seq. You suspect that the sex of the fish might be important so you track it too.

```{r, echo=FALSE, out.width='60%', fig.asp=.75, fig.align='center'}
knitr::include_graphics("images/week_01.005.jpeg")
```

## A biological example to get us started {.flexbox .vcenter}

- GETTING THE DATA READY TO ANALYZE 
    - How should the data set be organized to best analyze it? 
    - What are the key properties of the variables? 
    - Why does that matter for learning R?
    - Why does that matter for performing statistical analyses?

## {.flexbox .vcenter}

NOTE - Include a slide that highlights a data table for this experiment

## {.flexbox .vcenter}

```{r, echo=FALSE, out.width='100%', fig.asp=.75, fig.align='center'}
knitr::include_graphics("images/week_01.006.jpeg")
```

## Data set rules of thumb (aka Tidy Data) {.flexbox .vcenter}

- Store a copy of data in nonproprietary formats
- Leave an uncorrected file when doing analyses
- Maintain effective metadata about the data
- When you add observations to a database, add rows
- When you add variables to a database, add columns
- A column of data should contain only one data type

## {.flexbox .vcenter}

```{r, echo=FALSE, out.width='100%', fig.asp=.75, fig.align='center'}
knitr::include_graphics("images/week_01.007.jpeg")
```

## Why use `R`? {.flexbox .vcenter}

- Good general scripting tool for statistics and mathematics
- Powerful and flexible and free
- Runs on all computer platforms 
- New enhancements coming out all the time
- Superb data management & graphics capabilities
- Reproducibility - can keep your scripts to see exactly what was done
- You can write your own functions 
- Lots of online help available
- Can use a nice GUI front end such as `Rstudio`
- Can embed your `R` analyses in dynamic, polished files using `R markdown`

## `RMarkdown` {.flexbox .vcenter}

- A better way to embed code in R Markdown files
- http://kbroman.org/knitr_knutshell/pages/Rmarkdown.html
- You can insert `R chunks` into `Rmarkdown` documents
- Make your own `Rmarkdown` to perform the tasks today and later this week

## BASICS of `R` {.flexbox .vcenter .smaller}

- Commands can be submitted through the terminal, console or scripts
- In your scripts, anything that follows '#' symbol (aka hash) is just for humans
- Notice on these slides I'm evaluating the code chunks and showing output
- The output is shown here after the two `#` symbols and the number of output items is in `[]`
- Also notice that `R` follows the normal priority of mathematical evaluation


```{r basic multiplication, echo=TRUE}
4*4
```

```{r more multiplication, echo=TRUE}
(4+3*2^2)
```



## Assigning Variables {.flexbox .vcenter}
- A better way to do this is to assign variables
- Variables are assigned values using the `<-` operator.
- Variable names must begin with a letter, but other than that, just about anything goes. 
- Do keep in mind that `R` is case sensitive.

## Assigning Variables {.flexbox .vcenter}

```{r assigning variables, echo=TRUE, tidy=TRUE}

x <- 2
x*3
y <- x * 3
y-2

```

These do not work

```{r eval=FALSE, echo=TRUE}
3y <- 3
3*y <- 3
```


## Arithmetic operations on functions {}

- Arithmetic operations can be performed easily on functions as well as numbers.
- Try the following, and then your own.

```{r functions, eval = FALSE, echo = TRUE}
x+2
x^2
log(x)
```

- Note that the last of these - `log` - is a built in function of `R`, and therefore the object of the function needs to be put in parentheses
- These parentheses will be important, and we'll come back to them later when we add arguments after the object in the parentheses  
- The outcome of calculations can be assigned to new variables as well, and the results can be checked using the 'print' command


## Arithmetic operations on functions {.flexbox .vcenter}

```{r, echo=TRUE}
y <- 67
print(y)

x <- 124
z <- (x*y)^2
print(z)
```

## STRINGS {.flexbox .vcenter}
- Variables and operations can be performed on characters as well
- Note that characters need to be set off by quotation marks to differentiate them from numbers
- The `c` stands for `concatenate`
- Note that we are using the same variable names as we did previously, which means that we're overwriting our previous assignment
- A good rule of thumb is to use new names for each variable, and make them short but still descriptive

## STRINGS {.flexbox .vcenter}

```{r, echo=TRUE}
x <- "I Love"
print (x)
y <- "Biostatistics"
print (y)
z <- c(x,y)
print (z)
```

## FACTORS {.flexbox .vcenter}

- The variable `z` is now what is called a list of character values.
- Sometimes we would like to treat the characters as if they were units for subsequent calculations. 
- These are called `factors`, and we can redefine our character variables as factors.
- This might seem a bit strange, but itâ€™s important for statistical analyses where we might want to see the mean or variance for two different treatments.

## FACTORS {.flexbox .vcenter}

```{r, eval = FALSE, echo = TRUE}
z_factor <- as.factor(z)
print (z_factor)
```
- Note that factor levels are reported alphabetically

## VECTORS {.flexbox .vcenter}

- In general `R` thinks in terms of vectors (a list of characters, factors or numerical values) and it will benefit any `R` user to try to write programs with that in mind, as it will simplify most things.
- Vectors can be assigned directly using the 'c()' function and then entering the exact values.

## VECTORS {.flexbox .vcenter}

```{r Example vectors, echo=TRUE}
x <- c(2,3,4,2,1,2,4,5,10,8,9)
print(x)
```

## Basic Statistics {.flexbox .vcenter}

- Many functions exist to operate on vectors. 
- Combine these with your previous variable to see what happens. 
- Also, try to find other functions (e.g. standard deviation).

## Basic Statistics {.flexbox .vcenter}

```{r, eval = FALSE, echo = TRUE}
mean(x)
median(x)
var(x)
log(x)
ln(x)
sqrt(x)
sum(x)
length(x)
sample(x, replace = T)
```
- Notice that the last function (`sample`) has an argument (`replace=T`)
- Arguments simply modify or direct the function in some way
- There are many arguments for each function, some of which are defaults

## Getting Help {.flexbox .vcenter}

- Getting Help on any function is very easy - just type a question mark and the name of the function.
- There are functions for just about anything within `R` and it is easy enough to write your own functions if none already exist to do what you want to do.
- In general, function calls have a simple structure: a function name, a set of parentheses and an optional set of parameters to send to the function.
- Help pages exist for all functions that, at a minimum, explain what parameters exist for the function.  
- Help can be accessed a few ways - try them :

## Getting Help {.flexbox .vcenter}

```{r, eval = FALSE, echo = TRUE}
- help(mean)
- ?mean
- example(mean)
- help.search("mean")
- apropos("mean")
- args(mean)
```

## Creating vectors {.smaller}

- Creating vector of new data by entering it by hand can be a drag 
- However, it is also very easy to use functions such as `seq` and `sample`
- Try the examples below Can you figure out what the three arguments in the parentheses mean?
- Try varying the arguments to see what happens. 
- Don't go too crazy with the last one or your computer might slow way down

## Creating vectors {.smaller}

```{r, echo = TRUE}
seq_1 <- seq(0.0, 10.0, by = 0.1)
print(seq_1)
seq_2 <- seq(10.0, 0.0, by = -0.1)
print(seq_2)
```

## Creating vectors {.smaller}

```{r, echo = TRUE}
seq_square <- (seq_2)*(seq_2)
print(seq_square)
```

## Creating vectors {.smaller}

```{r, echo = TRUE}
seq_square_new <- (seq_2)^2
print(seq_square_new)
```


## Drawing samples from distributions {.flexbox .vcenter}

- Here is a way to create your own data sets that are random samples.
- Again, play around with the arguments in the parentheses to see what happens.

## Drawing samples from distributions {.smaller}

```{r Samples from distributions 1, out.width='100%', echo = TRUE}
x <- rnorm (10000, 0, 10)
y <- sample (1:10000, 10000, replace = T)
xy <- cbind(x,y)
plot(x,y) 
```

## Drawing samples from distributions {.smaller}

```{r Samples from distributions 2, out.width='100%', echo = TRUE}
x <- rnorm (10000, 0, 10)
y <- sample (1:10000, 10000, replace = T)
xy <- cbind(x,y)
plot(xy)
```

## Drawing samples from distributions {.smaller}

```{r Samples from distributions 3, out.width='100%', echo = TRUE}
x <- rnorm (10000, 0, 10)
y <- sample (1:10000, 10000, replace = T)
xy <- cbind(x,y)
hist(x)
```

## Drawing samples from distributions {.flexbox .vcenter}

- Youâ€™ve probably figured out that y from the last example is drawing numbers with equal probability.
- What if you want to draw from a distribution?
- Again, play around with the arguments in the parentheses to see what happens.

## Drawing samples from distributions {.flexbox .vcenter out.width='50%'}

```{r, out.width='100%', echo = TRUE}
x <-rnorm(1000, 0, 100)
hist(x, xlim = c(-500,500))
curve(50000*dnorm(x, 0, 100), xlim = c(-500,500), add=TRUE, col='Red')
```
- `dnorm()` generates the probability density, which can be plotted using the `curve()` function.
- Note that is curve is added to the plot using `add=TRUE` 

## Visualizing Data {.flexbox .vcenter}

- So far you've been visualizing just the list of output numbers
- Except for the last example where I snuck in a `hist` function.
- You can also visualize all of the variables that you've created using the `plot` function (as well as a number of more sophisticated plotting functions).
- Each of these is called a `high level` plotting function, which sets the stage
- `Low level` plotting functions will tweak the plots and make them beautiful

## Visualizing Data {.flexbox .vcenter}

- What do you think that each of the arguments means for the plot function?
- A cool thing about `R` is that the options for the arguments make sense. 
- Try adjusting an argument and see if it works
- Note next week we will be exploring the plotting in `GGPlot2`

## Visualizing Data {.flexbox .vcenter .smaller}

```{r, echo = TRUE}
seq_1 <- seq(0.0, 10.0, by = 0.1) 
plot (seq_1, xlab="space", ylab ="function of space", type = "p", col = "red")
```

## Putting plots in a single figure {.flexbox .vcenter}

- On the next slide
- The first line of the lower script tells R that you are going to create a composite figure that has two rows and two columns. Can you tell how?
- Now, modify the code to add two more variables and add one more row of two panels.

```{r, echo=TRUE, outwidth="50%", out.height="20%"}
seq_1 <- seq(0.0, 10.0, by = 0.1)
seq_2 <- seq(10.0, 0.0, by = -0.1)
```

## Putting plots in a single figure {.flexbox .vcenter .smaller}

```{r, echo=TRUE, outwidth="50%", out.height="50%"}
par(mfrow=c(2,2))
plot (seq_1, xlab="time", ylab ="p in population 1", type = "p", col = 'red')
plot (seq_2, xlab="time", ylab ="p in population 2", type = "p", col = 'green')
plot (seq_square, xlab="time", ylab ="p2 in population 2", type = "p", col = 'blue')
plot (seq_square_new, xlab="time", ylab ="p in population 1", type = "l", col = 'yellow')
```

## Example using binomial distribution {.flexbox .vcenter}

- As above for the normal distribution, data can be generated by being sampled from nearly any distribution and then visualized.
- Below Iâ€™m having you use the â€˜histogramâ€™ function. What does it do?

## Example using binomial distribution {.smaller}
- 10 successes (out of 20 trials) is the most frequent outcome

```{r binomial function, echo=TRUE}
heads <- rbinom(n=1000, size=20, prob=0.5)
hist(heads)
```

## Example using binomial distribution {.smaller}

- This kind of statement can be run in one line as well, which is sometimes easier.

```{r binomial function again, echo=TRUE}
hist(rbinom(n=1000, size=20, prob=0.5))
```

## Creating Data Frames in R {.smaller}

- As you have seen, in R you can generate your own random data set drawn from nearly any distribution very easily. 
- Often we will want to use collected data.
- Now, letâ€™s make a dummy dataset to get used to dealing with data frames
- Set up three variables (habitat, temp and elevation) as vectors

```{r, echo=TRUE}
habitat <- factor(c("mixed", "wet", "wet", "wet", "dry", "dry", "dry","mixed"))
temp <- c(3.4, 3.4, 8.4, 3, 5.6, 8.1, 8.3, 4.5)
elevation <- c(0, 9.2, 3.8, 5, 5.6, 4.1, 7.1, 5.3)
```

- Create a data frame where vectors become columns

```{r, echo=TRUE}
mydata <- data.frame(habitat, temp, elevation)
row.names(mydata) <- c("Reedy Lake", "Pearcadale", "Warneet", "Cranbourne", 
                       "Lysterfield", "Red Hill", "Devilbend", "Olinda")
```

- Now you have a hand-made data frame with row names

## Reading in Data Frames in R {.flexbox .vcenter}

- A strength of `R` is being able to import data from an external source
- Create the same table that you did above in a spreadsheet like Excel
- Export it to comma separated and tab separated text files for importing into `R`.
- The first will read in a comma-delimited file, whereas the second is a tab-delimited
- In both cases the header and row.names arguments indicate that there is a header row and row label column 
- Note that the name of the file by itself will have R look in the CWD, whereas a full path can also be used

## Reading in Data Frames in R {.flexbox .vcenter}

```{r, eval = FALSE, echo = TRUE}
YourFile <- read.table('yourfile.csv', header=T, row.names=1, sep=',')
YourFile <- read.table('yourfile.txt', header=T, row.names=1, sep='\t')
```

## Exporting Data Frames in R {.flexbox .vcenter}

```{r, eval = FALSE, echo = TRUE}
write.table(YourFile, "yourfile.csv", quote=F, row.names=T, sep=",")
write.table(YourFile, "yourfile.txt", quote=F, row.names=T, sep="\t")
```

## Indexing in data frames {.flexbox .vcenter}
- Next up - indexing just a subset of the data
- This is a very important idea in R, that you can analyze just a subset of the data.
- This is analyzing only the data in the file you made that has the factor value 'mixed'.

```{r, echo=TRUE, eval=FALSE}
print (YourFile[,2])
print (YourFile$temp)
print (YourFile[2,])
plot (YourFile$temp, YourFile$elevation)
```

## Indexing in data frames {.flexbox .vcenter}
- You can perform operations on particular levels of a factor
- Calculating the mean of the 'mixed' and 'gipps' levels of habitat. 
- Note that the first argument is the numerical column vector, and the second is the factor column vector. 
- The third is the operation. Reversing the first two does not work (the one below).

```{r, eval = FALSE, echo = TRUE}
tapply(YourFile$temp, YourFile$habitat, mean)
tapply(YourFile$temp, YourFile$habitat, var)
```

## R INTERLUDE | Some real transcriptomic data {.flexbox .vcenter}

- Examine the text file: `GacuRNAseq_Subset.csv`
- How many many rows and columns are there?
- How many different variables are there?
- What are the general types of variables?
- Now letâ€™s read the data file into R and analyze it
- This exercise will help you get used to reading in and manipulating genomic data files
- First off, remember to set your working directory to find your file correctly

## Some real transcriptomic data {.smaller}

```{r, eval = FALSE, echo = TRUE}
RNAseq_Data <- read.table('GacuRNAseq_subset.csv', header=TRUE, sep=',')

print (RNAseq_Data)
head (RNAseq_Data)
tail (RNAseq_Data)

print (RNAseq_Data[,2])
print (RNAseq_Data[1,])
print (RNAseq_Data[1,2])
print (RNAseq_Data$ENSGACG00000000010)
print (RNAseq_Data$ENSGACG00000000010>45.0)
```

## Summary stats and figures {.smaller}

```{r, eval = FALSE, echo = TRUE}
summary1 <- summary(RNAseq_Data $ENSGACG00000000003)
print (summary1)

hist(RNAseq_Data $ENSGACG00000000003)
boxplot(RNAseq_Data$ENSGACG00000000003)
boxplot(RNAseq_Data$ENSGACG00000000003~RNAseq_Data$Population)
plot(RNAseq_Data $ENSGACG00000000003, RNAseq_Data$ENSGACG00000000003)

boxplot(RNAseq_Data $ENSGACG00000000003~RNAseq_Data$Treatment, 
        col = "red", ylab = "Expression Level", xlab = "Treatment level", 
        border ="orange", 
        main = "Boxplot of variation in gene expression across microbiota treatments")
```

# Data wrangling and exploratory data analysis (EDA)

## Tidyverse family of packages {.smaller}

```{r, echo=FALSE, out.width='90%', fig.align='center'}
knitr::include_graphics("images/week_3.011.jpeg")
```

## Tidyverse family of packages {.vcenter .flexbox}

- Hadley Wickham and others have written R packages to modify data

- These packages do many of the same things as base functions in R

- However, they are specifically designed to do them faster and more easily

- Wickham also wrote the package GGPlot2 for elegant graphics creations

- GG stands for â€˜Grammar of Graphicsâ€™


## Example of a tibble {.smaller}

```{r, echo=FALSE, out.width='90%', fig.align='center'}
knitr::include_graphics("images/week_3.012.jpeg")
```

## Example of a tibble {.smaller}

```{r, echo=FALSE, out.width='90%', fig.align='center'}
knitr::include_graphics("images/week_3.013.jpeg")
```

## Types of vectors of data {.smaller}

```{r, echo=FALSE, out.width='50%', fig.align='center'}
knitr::include_graphics("images/week_3.014.jpeg")
```

## Types of vectors of data {.vcenter}

`int` stands for integers  
  
`db`l stands for doubles, or real numbers 
  
`chr` stands for character vectors, or strings  
  
`dttm` stands for date-times (a date + a time)  
  
`lgl` stands for logical, vectors that contain only TRUE or FALSE
  
`fctr` stands for factors, which R uses to represent categorical variables with fixed possible values  
  
`date` stands for dates  

## Types of vectors of data {.smaller}

- Logical vectors can take only three possible values: 
    - `FALSE`
    - `TRUE`
    - `NA` which is 'not available'.

- Integer and double vectors are known collectively as numeric vectors. 
    - In `R` numbers are doubles by default.

- Integers have one special value: NA, while doubles have four: 
    - `NA` 
    - `NaN` which is 'not a number'
    - `Inf`
    - `-Inf`
    
## Types of vectors of data {.smaller}

- R will also implicitly coerce the length of vectors. 
    - This is called **vector recycling**
    - The shorter vector is repeated or recycled
    - The shorter vector will be made the same length as the longer vector
    
- R will expand the shortest vector to the same length as the longest
    - This is so-called **recycling**
    - This is silent except when the length of the longer is not an integer multiple of the length of the shorter
    - When it is not youâ€™ll get an error

- The vectorised functions in tidyverse will throw errors when you recycle anything other than a scalar. 
    - If you do want to recycle something other than a scaler
    - Youâ€™ll need to do it yourself with `rep()`

## Key functions in `dplyr` for vectors 

- Pick observations by their values with `filter()`.
- Reorder the rows with `arrange()`.
- Pick variables by their names with `select()`.
- Create new variables with functions of existing variables with `mutate()`.
- Collapse many values down to a single summary with `summarise()`.

## `filter()`, `arrange()` & `select()`

```{r, echo=T, eval=F}
filter(flights, month == 11 | month == 12)
```

```{r, echo=T, eval=F}
arrange(flights, year, month, day)
```

```{r, echo=T, eval=F}
select(flights, year, month, day)
```

```{r, echo=FALSE, out.width='60%', fig.align='center'}
knitr::include_graphics("images/week_3.015.jpeg")
```


## `mutate()` & `transmutate()`

This function will add a new variable that is a function of other variable(s)

```{r, echo=T, eval=F}
mutate(flights_sml,
  gain = arr_delay - dep_delay,
  hours = air_time / 60,
  gain_per_hour = gain / hours
)
```

This function will eplace the old variable with the new variable

```{r, echo=T, eval=F}
transmute(flights,
  gain = arr_delay - dep_delay,
  hours = air_time / 60,
  gain_per_hour = gain / hours
)
```

## `group_by( )` & `summarize( )` {.smaller}

This first function allows you to aggregate data by values of categorical variables

```{r, echo=T, eval=F}
by_day <- group_by(flights, year, month, day)
```

Once you have done this aggregation, you can then calculate values (in this case the mean) of other variables split by the new aggregated levels of the categorical variable

```{r, echo=T, eval=F}
summarise(by_day, delay = mean(dep_delay, na.rm = TRUE))
```

- Note - you can get a lot of missing values! 
- Thatâ€™s because aggregation functions obey the usual rule of missing values: 
    - if thereâ€™s any missing value in the input, the output will be a missing value. 
    - fortunately, all aggregation functions have an na.rm argument which removes the missing values prior to computation

## R INTERLUDE | Playing with `Tidyverse` functions {.smaller}
- Examine my R script `TidyVerse.Râ€™`
- Step 1 - Read in the GacuRNAseq_Subset.csv dataset
- Step 2 - Make the dataset into a tibble
- Step 3 -  Select all of the categorical variables and only 4 of the gene count variables and put them into a new variable
- Step 4 - Mutate each of the 4 gene expression values by performing a square root transformation making a new variable for each of the original (keep all 8 in the dataset). 
- Step 5 - Summarize the mean and standard deviation for each of the gene count variables grouped by the â€˜sexâ€™ and â€˜populationâ€™ and â€˜treatmentâ€™ categorical variables
- Step 6 - Create a histogram for one of the original gene expression variables, and one of the derived variables  
- Step 7 - Create a box plot for one of the original gene expression variables, and one of the derived variables, split by treatment
- Step 8 - Write the final data table to a .csv file and one of the figures to a .pdf file 


# Git and GitHub

## Git and GitHub {.vcenter .flexbox}

https://learngitbranching.js.org/

-------

```{r, echo=FALSE, out.width='90%', fig.asp=.75, fig.align='center'}
knitr::include_graphics("images/images_5a.007.jpeg")
```

-------

```{r, echo=FALSE, out.width='90%', fig.asp=.75, fig.align='center'}
knitr::include_graphics("images/images_5a.008.jpeg")
```

-------

```{r, echo=FALSE, out.width='90%', fig.asp=.75, fig.align='center'}
knitr::include_graphics("images/images_5a.009.jpeg")
```

-------

```{r, echo=FALSE, out.width='90%', fig.asp=.75, fig.align='center'}
knitr::include_graphics("images/images_5a.010.jpeg")
```

-------

```{r, echo=FALSE, out.width='90%', fig.asp=.75, fig.align='center'}
knitr::include_graphics("images/images_5a.011.jpeg")
```

-------

```{r, echo=FALSE, out.width='90%', fig.asp=.75, fig.align='center'}
knitr::include_graphics("images/images_5a.012.jpeg")
```

-------

```{r, echo=FALSE, out.width='90%', fig.asp=.75, fig.align='center'}
knitr::include_graphics("images/images_5a.013.jpeg")
```

-------

```{r, echo=FALSE, out.width='90%', fig.asp=.75, fig.align='center'}
knitr::include_graphics("images/images_5a.014.jpeg")
```

## Clone the repository

- First make a new directory into which you will clone our course repository
- Open the terminal and navigate to the directory and type the following

```{bash, echo=TRUE, eval=FALSE}
git clone https://github.com/wcresko/UO_ABS.git
````

- Now to update the repository you just need to use these commands

```{bash, echo=TRUE, eval=FALSE}
git status

git merge origin/master
```
- The first command just tells you if anything has changed  
- If so, do the second!


